cdap-securestore-ext-gcp-secretstore/src/main/java/io/cdap/cdap/securestore/gcp/cloudsecretmanager/WrappedSecret.java

```
@@ -0,0 +1,117 @@
+   *
+   * @param resourceName Value to set for the "name" field needed for update operations.
+   */
+  public Secret getGcpSecret(String resourceName) {
+    return Secret.newBuilder()
+        // Set replication policy to automatic (as opposed to user-managed) and do not specify a
+        // CMEK (use google-managed key).
+        .setReplication(Replication.newBuilder().setAutomatic(Automatic.getDefaultInstance()))
+        .setName(resourceName)
+        .putAnnotations("cdap_namespace", namespace)
+        .putAnnotations("cdap_secret_name", secretMetadata.getName())
+        .putAnnotations(
+            "cdap_description", Optional.ofNullable(secretMetadata.getDescription()).orElse(""))
+        .putAnnotations("cdap_props", serializeProps(secretMetadata.getProperties()))
+        .build();
+  }
+
+  private static SecretMetadata toSecretMetadata(Secret secret) throws IOException {
+
+    return new SecretMetadata(
+        /* name= */ secret.getAnnotationsOrDefault("cdap_secret_name", ""),
+        /* description= */ secret.getAnnotationsOrDefault("cdap_description", ""),
+        /* creationTimeMs= */ Timestamps.toMillis(secret.getCreateTime()),
+        /* properties= */ deserializeProps(secret.getAnnotationsOrDefault("cdap_props", "{}")));
+  }
+
+  private static String getNamespace(Secret secret) {
+    return secret.getAnnotationsOrDefault("cdap_namespace", "");
+  }
+
+  private static String serializeProps(Map<String, String> input) {
+    return gson.toJson(input);
+  }
+
+  /**
+   * Parses {@code input} as JSON and returns the corresponding map, throwing an {@link IOException}
+   * if parsing input fails.
+   */
+  private static Map<String, String> deserializeProps(String input) throws IOException {
+    try {
+      return gson.fromJson(input, new TypeToken<Map<String, String>>() { }.getType());
+    } catch (JsonSyntaxException e) {
+      throw new IOException("Failed to parse cdap_props", e);
```

this should only happen if somebody messes with the property outside of this extension right?

I'm wondering if we want it to fail in that case, or whether it may make more sense to ignore it and return an empty map when that happens?

Another option could be to flatten the properties such that each key-value is a separate annotation (with some prefix like 'cdap_prop_'). That way we don't need to worry about malformed json strings.

---

Good point. Empty map might make more sense than throwing?

This is covered in the design doc so I'll leave this open and we can discuss offline.

---

let's keep this behavior where it fails on malformed values. Though I think it would be better to throw a more specific exception (ex: a new InvalidSecretException class) that indicates it's malformed rather than a generic IOException, as the caller may want to handle this case specifically.

For example, as written now, the list() call will fail if any of the secrets is malformed. It's probably better to skip malformed secrets and exclude them from the list rather than failing.